### DISCUSSION TITLE
About the information extraction in primal task

### DISCUSSION THREAD

**Discussion by superboom** [ds4dm/ml4co-competition#37] [Q&A] [answered]
1. In the primal task, is it feasible for us to include any open source MIP/LP solver, such as CBC,CLP and Soplex with their bin files?
2. if not, whether we can call the LP solver API in SCIP so that we can extract LP solution during the evaluating process?

**Comment by gasse** [comment, marked as answer]
Hi @superboom ,

> 1. In the primal task, is it feasible for us to include any open source MIP/LP solver, such as CBC, CLP and Soplex with their bin files?

Yes, as long as you find a way to call those from within the Python agent.

> 2. if not, whether we can call the LP solver API in SCIP so that we can extract LP solution during the evaluating process?

You can extract information from the internal model stored in the environment, from within the `ObservationFunction` of your agent. This includes the current LP solution value (`m.getLPObjVal()`). However, it is illegal to make any API call from PySCIPOpt that will change the state of the SCIP model, hence the state of the environment. You might consider making a *deep copy* of the SCIP model within your agent though, and then of course you can anything you want with it.
```python
m = model.as_pyscipopt()
m_copy = pyscipopt.Model(sourceModel=m)
```

Best,
Maxime
